### PR TITLE
Fix for: Invalid display name for links with mailto href

### LINKED ISSUE #2027: Invalid display name for links with mailto href
## Type of report

Bug

## Provide detailed reproduction steps (if any)

1. Open CKEditor sample with `link` plugin.
1. Add a link with display name `@foo` and href `foo@example.com`.
1. Click on the added link to open dialog.
1. Close dialog by clicking `ok` button.

linkmail

### Expected result

Link shouldn't change its display name.

### Actual result

Link has href as a display name.

## Other details

* Browser: Any
* OS: Any
* CKEditor version: 4.9.2
* Installed CKEditor plugins: link

### PR DESCRIPTION
**Pull request by Dumluregn**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What is the proposed changelog entry for this pull request?

`[(#2027)](https://github.com/ckeditor/ckeditor-dev/issues/2027): Fix for: Incorrect handling email display text starting with '@'.`

## What changes did you make?

Link plugin was changing email display text to be the same as email address every time it found '@' in display text. This condition was removed.

Closes #2027.

### CODE REVIEW

File: tests/plugins/link/emailencode.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	}
+};
+
+var protectedMailLink = '<a href=\"javascript:void(location.href=\'mailto:\'+String.fromCharCode(106,111,98,64,99,107,115,111,117,114,99,101,46,99,111,109)' +
+		'+\'?subject=Job%20Request&amp;body=I\\\'m%20looking%20for%20the%20AJD%20position.\')\">AJD</a>',
+	protectedMailLinkWithoutParams = '<a href=\"javascript:void(location.href=\'mailto:\'+' +
+		'String.fromCharCode(106,111,98,64,99,107,115,111,117,114,99,101,46,99,111,109))">AJD</a>';
+
+bender.test( {
+	'test created protected mail link': function() {
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply tests/plugins/link/emailencode.js]: Hi, this and other tests are just renamed and a bit cleaned up - they existed before and are not directly related to this PR. 
Also I just noticed I worked on them also in another PR (#2138), so I removed commits with duplicating changes from here.

File: tests/plugins/link/emailencode.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+			bodyField.setValue( 'I\'m looking for the AJD position.' );
+
+			dialog.fire( 'ok' );
+			dialog.hide();
+
+			assert.areSame( protectedMailLink, bot.getData( false, true ) );
+		} );
+	},
+
+	'test read from protected mail link': function() {
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply tests/plugins/link/emailencode.js]: ref https://github.com/ckeditor/ckeditor-dev/pull/3246#discussion_r312469316

File: tests/plugins/link/emailencode.html [hunk 3]
```diff
@@ -0,0 +1 @@
+<textarea id="test_create_link_editor"></textarea>
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply tests/plugins/link/emailencode.html]: ref https://github.com/ckeditor/ckeditor-dev/pull/3246#discussion_r312469316

File: tests/plugins/link/emailobfuscate.html [hunk 4]
```diff
@@ -0,0 +1 @@
+<textarea id="test_create_link_editor"></textarea>
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply tests/plugins/link/emailobfuscate.html]: ref https://github.com/ckeditor/ckeditor-dev/pull/3246#discussion_r312469316

File: tests/plugins/link/emailwithatsign.html [hunk 5]
```diff
@@ -0,0 +1 @@
+<textarea id="test_create_link_editor"></textarea>
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply tests/plugins/link/emailwithatsign.html]: ref https://github.com/ckeditor/ckeditor-dev/pull/3246#discussion_r312469316

File: tests/plugins/link/manual/emailwithatsign.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
+@bender-tags: bug, link, 4.13.0, 2027
+@bender-ui: collapsed
+@bender-ckeditor-plugins: link, toolbar, wysiwygarea, sourcearea, basicstyles, undo
+
+## Scenario 1:
+
+1. Open link dialog for first link (@CKSource).
+1. Click OK.
+
+**Expected:**
```
[the remark this answers is not in the document]
Comment (Dumluregn) [reply tests/plugins/link/manual/emailwithatsign.md]: Thanks, I struggled to find the exemplary styling for `md` files, is there any you are referencing to or it's just unwritten rule?

[the remark this answers is not in the document]
Comment (engineering-this) [reply tests/plugins/link/manual/emailwithatsign.md]: Unfortunately not, just unwritten rule 😞

**Comment by engineering-this** [comment]
I wonder about slightly changing how link dialog works. But this might need broader discussion.

When we create a link without filling `display text` field, then address is used as link text. However when we open dialog for just created link `display text` field is filled.

Screenshot 2019-08-22 at 14 45 50

Screenshot 2019-08-22 at 14 45 59

I think it'd be more convenient to not show anything in `display text`, so that dialog would look the same after opening it again.

This could drop logic for detecting if link changed, which causes the issue.

There would be two cases:
- Display text filled -> use it as link text
- Not filled -> use href as text

There would be no need to handle any extra cases when one of these fields were modified.

**Comment by f1ames** [comment]
TBH, after checking how `display text` works, it could use some refactoring as suggested in https://github.com/ckeditor/ckeditor-dev/pull/3246#issuecomment-523892465.

There are few issues now:
* Removing entire `display text` and confirming the dialog does nothing.
* Manually editing link protocol (in text) and then changing it in a dialog dropdown updates `display text` (which was earlier alter manually).

However, it needs some rethinking, @Dumluregn could you extract it to a separate issue?

**Comment by f1ames** [comment]
Rebased onto latest `major`.

**Comment by f1ames** [comment]
Rebased onto latest `master`.

**Pull request merged**